pkg/api/bzz.go

```diff
@@ -21,13 +28,255 @@ import (
 	"github.com/ethersphere/bee/pkg/file/loadsave"
 	"github.com/ethersphere/bee/pkg/jsonhttp"
 	"github.com/ethersphere/bee/pkg/manifest"
-	"github.com/ethersphere/bee/pkg/manifest/mantaray"
 	"github.com/ethersphere/bee/pkg/sctx"
 	"github.com/ethersphere/bee/pkg/storage"
 	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
 	"github.com/ethersphere/bee/pkg/tracing"
 )
 
+const (
+	multiPartFormData = "multipart/form-data"
+)
+
+func (s *server) bzzUploadHandler(w http.ResponseWriter, r *http.Request) {
+	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
+
+	contentType := r.Header.Get(contentTypeHeader)
+	mediaType, _, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		logger.Debugf("bzz upload: parse content type header %q: %v", contentType, err)
+		logger.Errorf("bzz upload: parse content type header %q", contentType)
+		jsonhttp.BadRequest(w, "invalid content-type header")
+		return
+	}
+	switch mediaType {
+	case contentTypeTar:
+		s.dirUploadHandler(w, r)
+	default:
+		s.fileUploadHandler(w, r)
+	}
+}
+
+// fileUploadResponse is returned when an HTTP request to upload a file is successful
+// type fileUploadResponse struct {
+// 	Reference swarm.Address `json:"reference"`
+// }
+
+// fileUploadHandler uploads the file and its metadata supplied as:
+// - multipart http message
+// - other content types as complete file body
+func (s *server) fileUploadHandler(w http.ResponseWriter, r *http.Request) {
+	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
+	var (
+		reader                  io.Reader
+		fileName, contentLength string
+		fileSize                uint64
+	)
+
+	// Content-Type has already been validated by this time
+	contentType := r.Header.Get(contentTypeHeader)
+	mediaType, params, _ := mime.ParseMediaType(contentType)
+
+	tag, created, err := s.getOrCreateTag(r.Header.Get(SwarmTagHeader))
+	if err != nil {
+		logger.Debugf("bzz upload file: get or create tag: %v", err)
+		logger.Error("bzz upload file: get or create tag")
+		jsonhttp.InternalServerError(w, "cannot get or create tag")
+		return
+	}
+
+	if !created {
+		// only in the case when tag is sent via header (i.e. not created by this request)
+		if estimatedTotalChunks := requestCalculateNumberOfChunks(r); estimatedTotalChunks > 0 {
+			err = tag.IncN(tags.TotalChunks, estimatedTotalChunks)
+			if err != nil {
+				s.logger.Debugf("bzz upload file: increment tag: %v", err)
+				s.logger.Error("bzz upload file: increment tag")
+				jsonhttp.InternalServerError(w, "increment tag")
+				return
+			}
+		}
+	}
+
+	// Add the tag to the context
+	ctx := sctx.SetTag(r.Context(), tag)
+
+	if mediaType == multiPartFormData {
+		mr := multipart.NewReader(r.Body, params["boundary"])
```

Comment:
i would remove the whole multipart section from the file upload part. cc @agazso